.docs/architecture/adr-001-general-parser-architecture.md

```diff
@@ -0,0 +1,227 @@
+# ADR-001: General parser architecture
+
+
+## Changelog
+
+- March 11th, 2024: Init draft;
+- March 13th, 2024: First draft;    
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+This ADR outlines the general parser architecture designed for Juno, addressing the need for enhanced interfaces to support various chains.
+
+## Context
+
+Juno currently lacks the necessary architecture documents and interfaces to effectively support multiple chains. Additionally, it faces challenges in handling messages and transactions, especially when dealing with different blockchain protocols and versions. There is a pressing need to establish a robust and adaptable architecture that can accommodate the diverse requirements of various chains while maintaining efficiency and reliability. 
+
+## Decision
+
+To address these challenges, the decision is made to redesign Juno's parser architecture. The architecture will prioritize modularity and extensibility, allowing seamless integration with different blockchain networks and protocols. Key components such as enqueuers, workers, and modules will be defined with clear responsibilities and interfaces to facilitate easy extension and customization. Moreover, the architecture will incorporate mechanisms to handle transactions, messages, and other blockchain operations efficiently.
+ 
+The provided image offer a comprehensive overview of the Juno parser, detailing its core components and functionalities.
+
+![Architecture](../.img/adr-001-general-architecture.png)
```

Comment:
I had something like this in mind.

<img width="2884" alt="Indexer" src="https://github.com/forbole/juno/assets/85519918/ecceb415-10ef-4560-acdb-87b1f1eb4f85">

Reply:
Yeah, it is great, we can make RabbitMQ or Kafka implement the `BlockQueue` interface if needed.

Reply:
Yeah, also I think that the Indexer and Parser part should be separated into their program/process. It will allow for less complexity in the code and some freedom in terms of the technologies used.

For instance, later we find that the Indexer wrote in Go has latency spikes due to garbage collection. Then it will be far easier to replace only that part with Haskell (just kidding obviously 😂). Another use-case would be if, I don't know, Bitcoin has SDKs written in C++, we would be able to write an Indexer just for that and re-use all the rest of the architecture.

Reply:
> Yeah, also I think that the Indexer and Parser part should be separated into their program/process. 

I agree with you. If we apply RabbitMQ or Kafka, we can make them sperately. That's why I starts from interfaces, and it brings us more possibilities to try and scale.

> For instance, later we find that the Indexer wrote in Go has latency spikes due to garbage collection. 

The Juno task would be Network IO bounding, most of time would be waiting for RPC server response, which takes about 200ms if Juno is not the same instance as RPC server. Therefore, gc spikes can be ignores in the case from my side.

> Another use-case would be if, I don't know, Bitcoin has SDKs written in C++, we would be able to write an Indexer just for that and re-use all the rest of the architecture.

We can use other languages if it is worth than Go, especially as you said, the native language of the chain provides SDK then we don't rewrite anything.